### ISSUE TITLE
Free-threaded builds fail

### ISSUE THREAD

**Issue by WhiteSymmetry** [hynek/argon2-cffi-bindings#67]
Successfully built jupyterlab
Failed to build argon2-cffi-bindings
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (argon2-cffi-bindings)

OS: Ubuntu 24.10 (Oracular Oriole)
gcc --version
gcc (Ubuntu 14.2.0-4ubuntu2) 14.2.0
Miniconda
Commands (mamba, conda, pip)
pip install jupyterlab
python -Im pip install --no-binary=argon2-cffi-bindings argon2-cffi-binding
python -Im pip install argon2-cffi-bindings

python -m pip install argon2-cffi-bindings
Collecting argon2-cffi-bindings
  Using cached argon2-cffi-bindings-21.2.0.tar.gz (1.8 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting cffi>=1.0.1 (from argon2-cffi-bindings)
  Using cached cffi-1.17.1-cp313-cp313t-linux_x86_64.whl
Collecting pycparser (from cffi>=1.0.1->argon2-cffi-bindings)
  Using cached pycparser-2.22-py3-none-any.whl.metadata (943 bytes)
Using cached pycparser-2.22-py3-none-any.whl (117 kB)
Building wheels for collected packages: argon2-cffi-bindings
  Building wheel for argon2-cffi-bindings (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for argon2-cffi-bindings (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [59 lines of output]
      <frozen importlib._bootstrap>:488: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module '_cffi_backend', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
      running bdist_wheel
      running build
      running build_py
      creating build/lib.linux-x86_64-cpython-313t/_argon2_cffi_bindings
      copying src/_argon2_cffi_bindings/__init__.py -> build/lib.linux-x86_64-cpython-313t/_argon2_cffi_bindings
      copying src/_argon2_cffi_bindings/_ffi_build.py -> build/lib.linux-x86_64-cpython-313t/_argon2_cffi_bindings
      running build_ext
      generating cffi module 'build/temp.linux-x86_64-cpython-313t/_ffi.c'
      creating build/temp.linux-x86_64-cpython-313t
      building '_ffi' extension
      creating build/temp.linux-x86_64-cpython-313t/build/temp.linux-x86_64-cpython-313t
      creating build/temp.linux-x86_64-cpython-313t/extras/libargon2/src
      creating build/temp.linux-x86_64-cpython-313t/extras/libargon2/src/blake2
      gcc -pthread -B /home/mk/miniconda3/envs/q13/shared/python_compiler_compat -fno-strict-overflow -Wsign-compare -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /home/mk/miniconda3/envs/q13/include -fPIC -O2 -isystem /home/mk/miniconda3/envs/q13/include -fPIC -Iextras/libargon2/include -I/home/mk/miniconda3/envs/q13/include/python3.13t -c build/temp.linux-x86_64-cpython-313t/_ffi.c -o build/temp.linux-x86_64-cpython-313t/build/temp.linux-x86_64-cpython-313t/_ffi.o -msse2
      In file included from build/temp.linux-x86_64-cpython-313t/_ffi.c:57:
      /home/mk/miniconda3/envs/q13/include/python3.13t/Python.h:51:4: error: #error "The limited API is not currently supported in the free-threaded build"
         51 | #  error "The limited API is not currently supported in the free-threaded build"
            |    ^~~~~
      In file included from /home/mk/miniconda3/envs/q13/include/python3.13t/Python.h:68:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:213:5: error: unknown type name ‘PyMutex’
        213 |     PyMutex ob_mutex;           // per-object lock
            |     ^~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘Py_REFCNT’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:314:22: error: implicit declaration of function ‘_Py_atomic_load_uint32_relaxed’ [-Wimplicit-function-declaration]
        314 |     uint32_t local = _Py_atomic_load_uint32_relaxed(&ob->ob_ref_local);
            |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:318:25: error: implicit declaration of function ‘_Py_atomic_load_ssize_relaxed’ [-Wimplicit-function-declaration]
        318 |     Py_ssize_t shared = _Py_atomic_load_ssize_relaxed(&ob->ob_ref_shared);
            |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘_Py_IsImmortal’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:352:63: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
        352 |     return (_Py_atomic_load_uint32_relaxed(&op->ob_ref_local) ==
            |                                                               ^~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘Py_SET_REFCNT’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:390:9: error: implicit declaration of function ‘_Py_IsOwnedByCurrentThread’ [-Wimplicit-function-declaration]
        390 |     if (_Py_IsOwnedByCurrentThread(ob)) {
            |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘Py_SET_SIZE’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:430:5: error: implicit declaration of function ‘_Py_atomic_store_ssize_relaxed’ [-Wimplicit-function-declaration]
        430 |     _Py_atomic_store_ssize_relaxed(&ob->ob_size, size);
            |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘Py_INCREF’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:811:9: error: implicit declaration of function ‘_Py_atomic_store_uint32_relaxed’ [-Wimplicit-function-declaration]
        811 |         _Py_atomic_store_uint32_relaxed(&op->ob_ref_local, new_local);
            |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:814:9: error: implicit declaration of function ‘_Py_atomic_add_ssize’; did you mean ‘__atomic_add_fetch’? [-Wimplicit-function-declaration]
        814 |         _Py_atomic_add_ssize(&op->ob_ref_shared, (1 << _Py_REF_SHARED_SHIFT));
            |         ^~~~~~~~~~~~~~~~~~~~
            |         __atomic_add_fetch
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h: In function ‘Py_DECREF’:
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:907:13: error: implicit declaration of function ‘_Py_MergeZeroLocalRefcount’ [-Wimplicit-function-declaration]
        907 |             _Py_MergeZeroLocalRefcount(op);
            |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/mk/miniconda3/envs/q13/include/python3.13t/object.h:911:9: error: implicit declaration of function ‘_Py_DecRefShared’; did you mean ‘_Py_DecRef’? [-Wimplicit-function-declaration]
        911 |         _Py_DecRefShared(op);
            |         ^~~~~~~~~~~~~~~~
            |         _Py_DecRef
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for argon2-cffi-bindings
Failed to build argon2-cffi-bindings
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (argon2-cffi-bindings)

**Comment by hynek** [comment]
What am I supposed to make of this? I don't know what you're trying to do, on what platform, using what commands. Your installer should be able to use argon2-cffi-bindings's wheels that work fine on all platforms that I have access to. Why it doesn't is not apparent from what you shared.

**Comment by hynek** [comment]
Thank you for the update – so it's Conda and you're attempting a free-threaded build?

It looks like those aren't supported by CFFI yet, so I don't think I can do anything about it for now: https://github.com/python-cffi/cffi/issues/126

**Comment by WhiteSymmetry** [comment]
pip install cffi
Collecting cffi
  Using cached cffi-1.17.1-cp313-cp313t-linux_x86_64.whl
Requirement already satisfied: pycparser in ./.local/lib/python3.13t/site-packages (from cffi) (2.22)
Installing collected packages: cffi
Successfully installed cffi-1.17.1

or

python3 -m pip install cffi
Defaulting to user installation because normal site-packages is not writeable
Collecting cffi
  Using cached cffi-1.17.1-cp313-cp313t-linux_x86_64.whl
Collecting pycparser (from cffi)
  Using cached pycparser-2.22-py3-none-any.whl.metadata (943 bytes)
Using cached pycparser-2.22-py3-none-any.whl (117 kB)
Installing collected packages: pycparser, cffi
Successfully installed cffi-1.17.1 pycparser-2.22

**Comment by WhiteSymmetry** [comment]
[linux-64/python-3.13.0rc2-h6355ac2_1_cp313t.conda](https://anaconda.org/conda-forge/python/3.13.0rc2/download/linux-64/python-3.13.0rc2-h6355ac2_1_cp313t.conda) In this version (Threading), it gives an installation error in this version.

[linux-64/python-3.13.0rc2-h9ebbce0_101_cp313.conda](https://anaconda.org/conda-forge/python/3.13.0rc2/download/linux-64/python-3.13.0rc2-h9ebbce0_101_cp313.conda) There is no problem with this version. 

python -m pip install argon2-cffi-bindings
Collecting argon2-cffi-bindings
  Downloading argon2_cffi_bindings-21.2.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)
Requirement already satisfied: cffi>=1.0.1 in ./miniconda3/envs/a13/lib/python3.13/site-packages (from argon2-cffi-bindings) (1.17.1)
Requirement already satisfied: pycparser in ./.local/lib/python3.13/site-packages (from cffi>=1.0.1->argon2-cffi-bindings) (2.22)
Downloading argon2_cffi_bindings-21.2.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (86 kB)
Installing collected packages: argon2-cffi-bindings
Successfully installed argon2-cffi-bindings-21.2.0

**Comment by tacaswell** [comment]
@minrk has a patch that builds:

https://github.com/hynek/argon2-cffi-bindings/compare/main...minrk:argon2-cffi-bindings:313t

I have been using this branch locally, but I'm not clear how I found it.

The downside is that with 313+jit that branch hangs in build.

**Comment by clin1234** [comment]
Similar error here:
```
Collecting argon2-cffi-bindings (from -r sys312.txt (line 9))
  Downloading argon2-cffi-bindings-21.2.0.tar.gz (1.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 33.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [64 lines of output]
      <frozen importlib._bootstrap>:488: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module '_cffi_backend', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
      running dist_info
      creating /tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info
      writing /tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/PKG-INFO
      writing dependency_links to /tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/dependency_links.txt
      writing requirements to /tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/requires.txt
      writing top-level names to /tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/top_level.txt
      writing manifest file '/tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/SOURCES.txt'
      ERROR setuptools_scm._file_finders.git listing git files failed - pretending there aren't any
      reading manifest file '/tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      writing manifest file '/tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings.egg-info/SOURCES.txt'
      creating '/tmp/pip-modern-metadata-rniv4ow2/argon2_cffi_bindings-21.2.0.dist-info'
      Traceback (most recent call last):
        File "/workspaces/venv/lib/python3.14t/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "/workspaces/venv/lib/python3.14t/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "/workspaces/venv/lib/python3.14t/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/build_meta.py", line 377, in prepare_metadata_for_build_wheel
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/build_meta.py", line 320, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 86, in <module>
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/__init__.py", line 117, in setup
          return distutils.core.setup(**attrs)
                 ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/core.py", line 183, in setup
          return run_commands(dist)
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/core.py", line 199, in run_commands
          dist.run_commands()
          ~~~~~~~~~~~~~~~~~^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/dist.py", line 954, in run_commands
          self.run_command(cmd)
          ~~~~~~~~~~~~~~~~^^^^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/dist.py", line 995, in run_command
          super().run_command(command)
          ~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/dist.py", line 973, in run_command
          cmd_obj.run()
          ~~~~~~~~~~~^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/command/dist_info.py", line 99, in run
          bdist_wheel = self.get_finalized_command('bdist_wheel')
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/cmd.py", line 303, in get_finalized_command
          cmd_obj.ensure_finalized()
          ~~~~~~~~~~~~~~~~~~~~~~~~^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/_distutils/cmd.py", line 111, in ensure_finalized
          self.finalize_options()
          ~~~~~~~~~~~~~~~~~~~~~^^
        File "<string>", line 75, in finalize_options
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/command/bdist_wheel.py", line 261, in finalize_options
          self._validate_py_limited_api()
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
        File "/tmp/pip-build-env-ze1wwqw1/overlay/lib/python3.14t/site-packages/setuptools/command/bdist_wheel.py", line 296, in _validate_py_limited_api
          raise ValueError(
          ...<4 lines>...
          )
      ValueError: `py_limited_api='cp314'` not supported. `Py_LIMITED_API` is currently incompatible with `Py_GIL_DISABLED` (sys.abiflags='t'). See https://github.com/python/cpython/issues/111506.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by stonebig** [comment]
Hope ?

**Comment by gaborbernat** [comment]
Yeah, same when trying to do this on tox, https://github.com/tox-dev/tox/actions/runs/14999479501/job/42142296713?pr=3526 but seems is because of https://github.com/python/cpython/issues/111506

**Comment by hynek** [comment]
So as an update, I got to talk to @ngoldbaum and @reaperhulk at length at PyCon US 2025 about the whole topic.

At the current state of the cffi project, there's really only three ways forward:

1. Someone manages to fix free-threading in cffi. This is something that is being actively worked on, but it's turned out to be far from trivial even if we accept the possibility of a community fork.
2. The cryptography project that has most of the APIs we need manages to get rid of its own dependency on CFFI. You might be surprised how a Rust-based project can depend on cffi, but you also might be surprised how much work goes into making it work as seamlessly as it does.

   Given they use a different Argon2 implementation (the one from OpenSSL), they also offer more APIs compared to the official argon2 bindings that haven't seen an update in 4 years. This would allow me to add secret peppering and associated data and retire this project as the main driver for argon2-cffi (goodbye Tidelift income 😢).

   Given also that it's difficult to have a non-trivial project in Python without depending on cryptography, this would be my favorite solution.
3. There seems to be a well-maintained Argon2 implementation in Rust: <https://docs.rs/argon2/>. I could probably wrap it using PyO3/Maturin inside of this project. The problem is that I've never done that before, even if I accept excluding exotic platforms (OTOH the old builds will go nowhere). This would have the upside of moving to a memory-safe language, but would be a lot more work for me (I forgot all the Rust I once had + there's a whole ecosystem to re-learn) and it feels a bit weird to depend on a crate that I have almost no control over (vs I know where to send donuts to bribe Alex and Paul).

In the meantime I would appreciate if we as a community could stop acting as if argon2-cffi is somehow a problem to be solved/dropped. cffi goes much deeper throughout our ecosystem and it's unlikely that nontrivial software doesn't run into at least one dependency somewhere.

**Comment by minrk** [comment]
depending on cryptography seems like a good idea, now that it has argon2. cryptography's APIs lack PHC serialization, as far as I can tell, so that would be nice to have, either keeping in this package or added to cryptography (https://github.com/pyca/cryptography/issues/12094). I fully understand waiting on CFFI as a dependency, and it's definitely not this project's problem/fault, because once cffi works there isn't much/anything to do here, and loads of projects need that work done, so that's where I think the focus should be, I think. It just happens that argon2-cffi is the direct cffi dependency of some popular projects (particularly Jupyter Server), so it's the name that pip shows errors next to when CFFI is really the heart of it. I fully support directing folks to the CFFI efforts from here and saying there's nothing actionable on this repo until there's a usable solution for cffi, and this repo can only really "wait and see".

From the Jupyter side, it's tantalizing that argon2-cffi _happens_ to be the only hard cffi dependency in the Jupyter core stack (this surprised me) and it's used for just one optional function, so it's attractive to possibly shift dependencies around to unblock folks from trying out free-threading. Especially since scrypt is in the standard library now (though [PHC](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md) serialization is not) and totally adequate for Jupyter's purposes. I had started [exploring](https://github.com/jupyter-server/jupyter_server/pull/1520) making argon2-cffi optional, but of course ran into the fact that our _test_ dependencies also depend on CFFI (via hatch).

I am equally happy from the Jupyter side supporting scrypt and shifting argon2-cffi to optional to take some pressure off here, _or_ keeping it required if you'd actually like to keep that pressure on CFFI being a blocker for free-threaded installs of Jupyter (alongside hatch and loads of other things that depend directly or transitively on CFFI). Whatever feels most supportive to you is really fine by me. I don't personally feel great pressure to unblock free-threaded installs of Jupyter Server, especially since folks can install free-threaded _kernels_ already, and that's where it's likely to be most valuable.

**Comment by ngoldbaum** [comment]
For what it's worth I'm very aware cffi is a big pain point across the ecosystem and I'm actively working on getting our fork ready.

Unfortunately there's some complicated C coding that needs to happen to isolate mutable and immutable state and then add atomics and/or locking to the lazy initialization that cffi uses (and argon2-cffi relies on).

Hopefully I'll have something more concrete to share soon.

**Comment by hynek** [comment]
> depending on cryptography seems like a good idea, now that it has argon2. cryptography's APIs lack PHC serialization, as far as I can tell, so that would be nice to have, either keeping in this package or added to cryptography ([pyca/cryptography#12094](https://github.com/pyca/cryptography/issues/12094)).

They seem to have added it in https://github.com/pyca/cryptography/commit/c93ee3839379c109f17eea895796dc39415dc45c#diff-944fd2c9f14cea5d8b8e9ecaa80bc5c63c0e7720da2f6979f894813b61ec299b, no?

Either way, argon2-cffi ships an own PHC "parser" in https://github.com/hynek/argon2-cffi/blob/bc861f04ffa8a39e5c27fe963dc9b8815564c39b/src/argon2/_utils.py#L112-L159, so that isn't a factor either way.

AFAICT the only problem here is the lack of support for I and D schemes that are relevant in non-password contexts.

And that they depend on cffi too lol: https://github.com/pyca/cryptography/issues/12489#issuecomment-2755552891 (cursed be pyOpenSSL).

In other words, there's still a bunch yaks to shave on this route, too. :|

> From the Jupyter side, it's tantalizing that argon2-cffi _happens_ to be the only hard cffi dependency in the Jupyter core stack (this surprised me) and it's used for just one optional function, so it's attractive to possibly shift dependencies around to unblock folks from trying out free-threading. Especially since scrypt is in the standard library now (though [PHC](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md) serialization is not) and totally adequate for Jupyter's purposes. I had started [exploring](https://github.com/jupyter-server/jupyter_server/pull/1520) making argon2-cffi optional, but of course ran into the fact that our _test_ dependencies also depend on CFFI (via hatch).

Yeah, there's turtles all the way down. I guess we can be grateful it's _just_ CFFI that is causing so much trouble for now.

I guess the best solution for _this_ project would be Maturin, but I suspect there's a bunch of unknown unknowns waiting for me.

> I am equally happy from the Jupyter side supporting scrypt and shifting argon2-cffi to optional to take some pressure off here, _or_ keeping it required if you'd actually like to keep that pressure on CFFI being a blocker for free-threaded installs of Jupyter (alongside hatch and loads of other things that depend directly or transitively on CFFI). Whatever feels most supportive to you is really fine by me.

Well, I would prefer to not be made redundant if it doesn't help anyway. 🤓

> I don't personally feel great pressure to unblock free-threaded installs of Jupyter Server, especially since folks can install free-threaded _kernels_ already, and that's where it's likely to be most valuable.

Yeah, as excited about free-threading I am, it seems like it wouldn't be super wise to make rash moves and it's better to try to find paths forward that help the whole community instead of just cutting of projects left and right because they don't move fast enough. I'm fairly sure there's gonna be more blockers as we go.

**Comment by minrk** [comment]
> Well, I would prefer to not be made redundant if it doesn't help anyway. 🤓

Sure! I'll skip my effort to make it optional, then. To be clear since I didn't state it here, the scope of https://github.com/jupyter-server/jupyter_server/pull/1520 is only to make it optional on free-threaded where and while it doesn't install (using py>=3.13 as an imprecise proxy for free-threaded while we wait for PEP 780), and putting it back as a required dep where/when it works. So only ever a temporary measure.

I don't personally think argon2-cffi needs to be taking any particular pressing action other than waiting/watching cffi, unless _you_ feel the need or desire to do so. Waiting for cffi feels AOK to me.
